### PR TITLE
HTBHF-2012 Sending report payment message when making a payment.

### PR DESCRIPTION
**Pull request by dwybourn**
- Now sending a report payment message when making an initial payment, scheduled payment and top up payment.

### CODE REVIEW

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        ReportPaymentMessagePayload payload = createPayloadForNonTopUpPayments(claim, paymentCycle, INITIAL_PAYMENT);
+        messageQueueClient.sendMessage(payload, REPORT_PAYMENT);
+    }
+
+    public void sendReportScheduledPayment(Claim claim, PaymentCycle paymentCycle) {
+        ReportPaymentMessagePayload payload = createPayloadForNonTopUpPayments(claim, paymentCycle, SCHEDULED_PAYMENT);
+        messageQueueClient.sendMessage(payload, REPORT_PAYMENT);
+    }
+
+    private ReportPaymentMessagePayload createPayloadForNonTopUpPayments(Claim claim, PaymentCycle paymentCycle, PaymentAction paymentAction) {
```
Comment (dwybourn) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java]: Not too happy with the name here but couldn't think of a better one, open to suggestions.

Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java]: `createPayloadForPaymentCycle`?

Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java]: Changed to createPayloadWithPaymentAmountsDerivedFromPaymentCycle.

**Comment by dwybourn** [comment]
> I think we're missing a check for this report in the PaymentCycleIntegrationTest, but that could go in a new PR

The final PR is integration tests only and includes an update for PaymentCycleIntegrationTest

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/AdditionalPregnancyPaymentMessageProcessor.java [hunk 2]
```diff
@@ -59,6 +63,7 @@ public MessageStatus processMessage(Message message) {
         int paymentAmountInPence = calculatePaymentAmountInPence(message, context, paymentCycle);
         if (paymentAmountInPence > 0) {
             paymentService.makeInterimPayment(paymentCycle.get(), context.getClaim().getCardAccountId(), paymentAmountInPence);
+            reportPaymentMessageSender.sendReportTopUpPaymentMessage(paymentCycle.get().getClaim(), paymentCycle.get(), paymentAmountInPence);
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/AdditionalPregnancyPaymentMessageProcessor.java]: I think as this is going to be specific to pregnancy payments, the method should be renamed accordingly to be pregnancy specific. If its going to be more generic, then it should be moved into the makeInterimPayment method. Sounds like the former is correct.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java [hunk 3]
```diff
@@ -68,6 +70,9 @@
     @MockBean
     private PaymentCalculator paymentCalculator;
 
+    @SpyBean
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java]: A MockBean here is sufficient as you don't want/need the code to actually be called.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/AdditionalPregnancyPaymentMessageProcessor.java (hunk 2 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/AdditionalPregnancyPaymentMessageProcessor.java]: Renamed to sendReportPregnancyTopUpPaymentMessage().

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java (hunk 3 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java]: Replaced with a MockBean.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * Class responsible for sending report payment messages.
+ */
+@Component
+@AllArgsConstructor
+public class ReportPaymentMessageSender {
+
+    private final MessageQueueClient messageQueueClient;
+
+    /**
+     * Sends a report payment message with a {@link PaymentAction} of {@link PaymentAction#TOP_UP_PAYMENT}.
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java]: Minor - might be worth mentioning pregnancy in the JavaDoc.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    public void sendReportInitialPaymentMessage(Claim claim, PaymentCycle paymentCycle) {
+        ReportPaymentMessagePayload payload = createPayloadWithPaymentAmountsDerivedFromPaymentCycle(claim, paymentCycle, INITIAL_PAYMENT);
+        messageQueueClient.sendMessage(payload, REPORT_PAYMENT);
+    }
+
+    public void sendReportScheduledPayment(Claim claim, PaymentCycle paymentCycle) {
+        ReportPaymentMessagePayload payload = createPayloadWithPaymentAmountsDerivedFromPaymentCycle(claim, paymentCycle, SCHEDULED_PAYMENT);
+        messageQueueClient.sendMessage(payload, REPORT_PAYMENT);
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java]: Minor - should we have a private method for this line as its repeated 3 times?

Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/ReportPaymentMessageSender.java]: Will leave as is as it reads quite nicely and is only one line.

**Pull request merged**